reference/JSDoc-Type-References.md

```diff
@@ -0,0 +1,242 @@
+In Javascript files, the compiler understands Typescript types in
+JSDoc comments like `@type` and `@param`. However, it has additional
+rules to make type resolution work better for Javascript code that
+wasn't written with machine checking in mind. These rules apply only
+in JSDoc in Javascript files, and some can be disabled by setting
+`"noImplicitAny": true`, which the compiler takes as a signal that the
+code is being written with Typescript in mind.
+
+This document first describes the way Javascript type resolution is
+different, because those differences are bigger and more surprising.
+Then it covers differences in name resolution.
+
+## Simple Rewrites ##
+
+The simplest Javascript-only rule is to rewrite the primitive object
+types to real primitives.
+
+Type        | Resolved Type
+------------|--------------
+`Number`    | `number`
+`String`    | `string`
+`Boolean`   | `boolean`
+`Object`    | `any`
+
+The rewrite `Object -> any` is disabled when `"noImplicitAny": true`,
+which the compiler takes as a signal that the code is being written
+with Typescript in mind. Old versions of Typescript also rewrote
+`object -> any`, but stopped in Typescript 3.7.
+
+There are also rewrites of other JSDoc types that are equivalent to
+built-in Typescript types:
+
+Type                  | Resolved Type
+----------------------|--------------
+`Null`                | `null`
+`Undefined`           | `undefined`
+`Void`                | `void`
+`function`            | `Function`
+`Object<string,Type>` | `{ [x: string]: Type }`
+`Object<number,Type>` | `{ [x: number]: Type }`
+
+Finally, `Array` and `Promise` both default their type parameters to
+`any` when no type parameters are provided:
+
+Type        | Resolved Type
+------------|--------------
+`Array`     | `any[]`
+`Promise`   | `Promise<any>`
+
+This is standard for all type references in JSDoc, so I don't think
+it's needed anymore. It only applies when `"noImplicitAny": false`.
+
+### Where To Find The Code ###
+
+`getIntendedJSDocTypeReference` in src/compiler/checker.ts.
+
+## Values as Types ##
```

Comment:
I'll be honest, I'd prefer to never mention this in any documentation ever so that we don't imply that we think relying on it is a good idea (it's not).

Reply:
Like maybe we could talk about it in the TypeScript necronomicon or something, but I'd leave it out of the handbook.

Reply:
Reference pages *are* the necronomicon, or the closest we'll get in an open-source team.

This is why I've never written it down before, though. I hoped it would "just work" for most people.

Reply:
But, like, `rust` has a hard split between "The Rust Book" and "The Rustonomicon" (which is what I was referencing before). While documenting the current state of esoteric edge cases in, say, conditional type resolution or js type lookup is laudable, I don't think it should be adjacent to information that's much more mainstream - it serves mostly to distract. Like, I think the "what simple remappings do we do (`String -> string`)" is useful for casual reading, while "what fallback code exists within the compiler, approximately" is heavyweight "I wanna know how the compiler internals tick" stuff. This article does a bit of both - I'd split it into down, down those viewership lines, tbh.

Reply:
I think we're going for the same two categories, but with a softer split. Look at the other documents in reference/ -- they cover the same kinds of things.

I think it is worthwhile to put the first section's information into the handbook under checkJs though.
It's just that that chapter doesn't exist yet.

Reply:
Does that mean that `Compiler Options` belongs elsewhere? Because that _is_ just a reference, while the rest are really technical internal discussions.

Reply:
Probably. Not sure where it *should* go, though. @RyanCavanaugh opinions?

Reply:
Yeah, I think this is good case for splitting into two articles - my goal when I get my hands on doing new handbook stuff is to try split the existing doc into three main camps:

- Language Learners
- Experts looking to debug why something isn't what they expect
- People wanting to faff with tooling

And each doc on any topic should ideally only be trying to address one of those three people (e.g. don't try mix the descriptions of config settings deeply in with a language feature) 

I expected we'd need a bunch of docs for JSDoc though, so I wouldn't worry too much now about how it splits because once it starts becoming public facing we'll have more info

Reply:
@orta when you say "faff with tooling", are you thinking of API consumers? This article is mostly people interested in compiler internals, with a touch of "experts' debugging help". That's what I was thinking of for the Assignability page too; realistically knowing the mechanics of assignability isn't going to help debugging much.

Reply:
For example "TypeScript with webpack" or "TypeScript with eslint" user-level stuff - not this depth 👍